ml/utils.py

```diff
@@ -0,0 +1,354 @@
+    # y is a pandas series of labels for a given dataset
+    sorted_values = y.sort_values().unique()
+    similarity_encoder = SimilarityEncoder(similarity="ngram")
+    transformed_values = similarity_encoder.fit_transform(sorted_values.reshape(-1, 1))
+
+    plot_MDS(transformed_values, similarity_encoder, sorted_values, n_points=n_points)
+    return transformed_values, similarity_encoder, sorted_values
+
+
+def plot_MDS(transformed_values, similarity_encoder, sorted_values, n_points=15):
+    # TODO try this with UMAP
+    mds = MDS(dissimilarity="precomputed", n_init=10, random_state=42)
+    two_dim_data = mds.fit_transform(1 - transformed_values)  # transformed values lie
+
+    random_points = np.random.choice(
+        len(similarity_encoder.categories_[0]), n_points, replace=False
+    )
+
+    nn = NearestNeighbors(n_neighbors=2).fit(transformed_values)
+    _, indices_ = nn.kneighbors(transformed_values[random_points])
+    indices = np.unique(indices_.squeeze())
+
+    f, ax = plt.subplots()
+    ax.scatter(x=two_dim_data[indices, 0], y=two_dim_data[indices, 1])
+    # adding the legend
+    for x in indices:
+        ax.text(
+            x=two_dim_data[x, 0], y=two_dim_data[x, 1], s=sorted_values[x], fontsize=8
+        )
+    ax.set_title(
+        "multi-dimensional-scaling representation using a 3gram similarity matrix"
+    )
+
+    f2, ax2 = plt.subplots(figsize=(7, 7))
+    cax2 = ax2.matshow(transformed_values[indices, :][:, indices])
+    ax2.set_yticks(np.arange(len(indices)))
+    ax2.set_xticks(np.arange(len(indices)))
+    ax2.set_yticklabels(sorted_values[indices], rotation="30")
+    ax2.set_xticklabels(sorted_values[indices], rotation="60", ha="right")
+    ax2.xaxis.tick_bottom()
+    ax2.set_title("Similarities across categories")
+    f2.colorbar(cax2)
+    f2.tight_layout()
+
+
+# #################################################################################################
+#
+#   Finding subgraphs within a large graph.
+#
+# #################################################################################################
+
+
+def search_to_df(word, col, df):
+    """
+    A simple way to retrieve entries from a given col in a dataframe
+    :eg
+        search_to_df('BlackRock', 'to_node', edf)
+
+    :param word: str search term
+    :param col: given column of dataframe
+    :param df: pandas dataframe
+    :returns
+        DataFrame of results
+    """
+    word = word.lower()
+    try:
+        res = df[df[col].apply(lambda x: word in str(x).lower())]
+    except TypeError as e:
+        logger.error(e)
+        return df
+    return res
+
+
+def get_nearest(search_term, src, dst, edf):
+    """
+    :param search_term:
+    :param src:
+    :param dst:
+    :param edf:
+    :return: pandas.DataFrame
+    """
+    logger.info(f"Finding {search_term} in both {src} and {dst} columns")
+    tdf = pd.concat(
+        [search_to_df(search_term, src, edf), search_to_df(search_term, dst, edf)],
```

Comment:
... is this right?

it seems like some ~NLP helpers for pattern search are getting added (cypher-lite)... we may want to keep as internal pygraphistry api for now as likely unstable